### PR TITLE
refactor(cri): restructure CRI API (improve robustness, clarity and maintainability)

### PR DESCRIPTION
**Pull request by incertum**
**What type of PR is this?**

> Uncomment one (or more) `/kind <>` lines:

> /kind bug

/kind cleanup

> /kind design

> /kind documentation

> /kind failing-test

> /kind feature

**Any specific area of the project related to this PR?**

> Uncomment one (or more) `/area <>` lines:

> /area API-version

> /area build

> /area CI

> /area driver-kmod

> /area driver-bpf

> /area driver-modern-bpf

> /area libscap-engine-bpf

> /area libscap-engine-gvisor

> /area libscap-engine-kmod

> /area libscap-engine-modern-bpf

> /area libscap-engine-nodriver

> /area libscap-engine-noop

> /area libscap-engine-source-plugin

> /area libscap-engine-savefile

> /area libscap

> /area libpman

/area libsinsp

> /area tests

> /area proposals

**Does this PR require a change in the driver versions?**

> /version driver-API-version-major

> /version driver-API-version-minor

> /version driver-API-version-patch

> /version driver-SCHEMA-version-major

> /version driver-SCHEMA-version-minor

> /version driver-SCHEMA-version-patch

**What this PR does / why we need it**:

I have added very detailed comments to the code diff re the implemented changes.

At a high level the refactor achieves the following:

- Add robustness wrt to metadata retrieved from the PodSandboxStatusResponse
- Improve technical clarity and maintainability (flow and steps are now easy to follow and understand)
- Remove some redundancies
- Expand unit tests (further improving technical clarity and maintainability)
- This PR does NOT add any extra API calls that weren't done prior to this refactor

**Which issue(s) this PR fixes**:

https://github.com/falcosecurity/libs/issues/1589

Why this PR matters:
- Currently, we are not handling or enriching pod sandbox container processes correctly. This means adopters can falsely report that a container image is missing, when in fact it may be legitimate. Once this PR is merged, adopters will have the option to compare container and pod sandbox IDs. If they are equal, they will know that the container images will never exist.
- Additionally, I would rate this code cleanup as highly crucial for slightly optimizing the CRI API lookups. More importantly, it will make them unit-testable and improve transparency and maintainability in the long run.

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

File: userspace/libsinsp/cri.hpp
Comment (incertum) [inline userspace/libsinsp/cri.hpp:660]: Removing `parse_containerd` brings significant benefits by enhancing the clarity and transparency of the codebase. This is because the parsing previously done here always also applied for `cri-o` and therefore at best this approach was confusing overall.

Now the main `parse` function calls all parsers and we get rid of this nested additional jump.

Also I have the suspicion we were inefficient with the IP enrichment / lookup before (it may have been called twice?), now the main`parse` function has a very clear structure ensuring we make only one call for container and pod sandbox status responses and parse them right away.

File: userspace/libsinsp/cri.hpp
Comment (incertum) [inline userspace/libsinsp/cri.hpp:725]: We previously didn't do a good job in handling pod sandbox containers and the labels (container or pod sandbox labels) were confusing in any aspect. Now each container regardless of whether it is a normal container or a pod sandbox container has `m_labels` and `m_pod_sandbox_labels`. 

Further improvements include adding all info that makes sense. Now pod sandbox container processes are also enriched the best way possible. Note that in the k8s filterchecks we do not need pod sandbox containers anymore to retrieve `m_pod_sandbox_labels`.

File: userspace/libsinsp/cri.hpp
Comment (incertum) [inline userspace/libsinsp/cri.hpp:750]: No more ad-hoc code here, the main `parse` function consistently calls `parse_*` helpers.

File: userspace/libsinsp/cri.hpp
Comment (incertum) [inline userspace/libsinsp/cri.hpp:776]: todo confirm

File: userspace/libsinsp/cri.hpp
Comment (incertum) [inline userspace/libsinsp/cri.hpp:793]: I mentioned before that we may have done the network enrichment twice before and now only one time in a transparent and clean way.

File: userspace/libsinsp/sinsp_filtercheck_k8s.cpp
Comment (incertum) [inline userspace/libsinsp/sinsp_filtercheck_k8s.cpp:230]: @leogr @Andreagit97 this should satisfy the desired behavior, please confirm.

File: userspace/libsinsp/sinsp_filtercheck_k8s.cpp
Comment (incertum) [inline userspace/libsinsp/sinsp_filtercheck_k8s.cpp:286]: No more extra pod sandbox container lookups from the container cache (now more efficient). Plus I suspect this may have caused inconsistencies in corner cases as those containers were processed at different times in the async event queue.

File: userspace/libsinsp/sinsp_filtercheck_k8s.cpp
Comment (incertum) [inline userspace/libsinsp/sinsp_filtercheck_k8s.cpp:304]: Clear docs.

File: userspace/libsinsp/test/container_engine/container_parser_cri_containerd.ut.cpp
Comment (incertum) [inline userspace/libsinsp/test/container_engine/container_parser_cri_containerd.ut.cpp:56]: only adjusted indentation

File: userspace/libsinsp/test/container_engine/container_parser_cri_containerd.ut.cpp
Comment (incertum) [inline userspace/libsinsp/test/container_engine/container_parser_cri_containerd.ut.cpp:694]: Now extra tests for pod sandbox containers.

File: userspace/libsinsp/test/filterchecks/k8s.cpp
Comment (incertum) [inline userspace/libsinsp/test/filterchecks/k8s.cpp:201]: This test now also clearly shows that we have all we need in the container, no need for extra pod sandbox container lookup from the container cache anymore.

File: userspace/libsinsp/cri.hpp
Comment (incertum) [reply userspace/libsinsp/cri.hpp:793]: @gnosek highlighting again that I am not introducing extra CRI API calls. They were already there before. 

In summary, the design is that for each container that is not a pod sandbox container you get a `ContainerStatsResponse` once and also a `PodSandboxStatusResponse` once and fully make use of the fact that you have the `PodSandboxStatusResponse`.

If we want to complicate things and cache `PodSandboxStatusResponse` info in pod sandbox containers so that the call is only done once for each pod is a different discussion. I would not recommend complicating things for now, but we could think about it in the future.

File: userspace/libsinsp/cri.hpp
Comment (incertum) [inline userspace/libsinsp/cri.hpp:577]: Let's double check if we can adjust this check for clear behavior ....

File: userspace/libsinsp/cri.hpp
Comment (incertum) [inline userspace/libsinsp/cri.hpp:790]: @gnosek highlighting again that I am not introducing extra CRI API calls. They were already there before. 

In summary, the design is that for each container that is not a pod sandbox container you get a `ContainerStatsResponse` once and also a `PodSandboxStatusResponse` once and fully make use of the fact that you have the `PodSandboxStatusResponse`.

If we want to complicate things and cache `PodSandboxStatusResponse` info in pod sandbox containers so that the call is only done once for each pod is a different discussion. I would not recommend complicating things for now, but we could think about it in the future.

File: userspace/libsinsp/sinsp_filtercheck_k8s.cpp
Comment (incertum) [inline userspace/libsinsp/sinsp_filtercheck_k8s.cpp:285]: For maximum robustness in retrieving the fields we get from the PodSandboxStatusResponse fetch them either from the container or as fallback from the pod sandbox container.

**Comment by incertum** [comment]
@gnosek may I ask for a first-pass review from you? Thanks a bunch in advance?

**Comment by incertum** [comment]
If ok with everyone

/milestone 0.15.0

**Comment by incertum** [comment]
@falcosecurity/libs-maintainers kindly checking in on the status of this PR. Thank you.

**Comment by incertum** [comment]
Please note that this PR has been open for 2 months now without receiving any review whatsoever. During this time, it appears that our focus has been primarily directed towards prioritizing and integrating similarly complex code alterations. 

Why this PR matters:
- Currently, we are not handling or enriching pod sandbox container processes correctly. This means adopters can falsely report that a container image is missing, when in fact it may be legitimate. Once this PR is merged, adopters will have the option to compare container and pod sandbox IDs. If they are equal, they will know that the container images will never exist.
- Additionally, I would rate this code cleanup as highly crucial for slightly optimizing the CRI API lookups. More importantly, it will make them unit-testable and improve transparency and maintainability in the long run.

**Comment by FedeDP** [comment]
PR LGTM, but i am in no way an expert on this part of the code :) 
I really like the refactor though, that is the most important thing.
/cc @gnosek :pray:

**Comment by leogr** [comment]
On my side, I will prioritize this once https://github.com/falcosecurity/libs/pull/1595 gets merged (hopefully soon).
/assign

I still want to confirm that this is targetted for libs 0.15

**Comment by incertum** [comment]
@leogr this PR should be the next one. Rebased and added another commit on top in an attempt to improve comments and naming and code organization even more.

If it help, happy to schedule a call to more interactively review this PR, just let me know, thanks a bunch!

**Comment by leogr** [comment]
> @leogr this PR should be the next one. Rebased and added another commit on top in an attempt to improve comments and naming and code organization even more.
> 
> If it help, happy to schedule a call to more interactively review this PR, just let me know, thanks a bunch!

I'm working on it. @therealbobo already did some tests, and everything worked properly (@therealbobo do you confirm?)/

cc @falcosecurity/libs-maintainers for visibility

**Comment by incertum** [comment]
> I'm working on it. @therealbobo already did some tests, and everything worked properly (@therealbobo do you confirm?)/

Thank you! While it's a terrible diff, I believe this PR is less risky, plus at least it's backed up my unit tests (acknowledging some limitations w/ mock CRI response unit tests, but still better than before when we had nothing).

Perhaps, double check and pay close attention to all conditional checks etc. 

Plus plz let me know if the filterchecks fallback mechanisms are ok with you:
- We now consistently first try to get all pod sandbox related fields from the container
- Falling back to fetching the separate pod sandbox container from the cache (which is now also fully populated, it was not the case before). The `cniresult` string was chosen as check for it as it should always be there in Kubernetes, while labels could legitimately be empty who knows. perhaps down the road there could be another improvement PR< but I wouldn't aim for absolutely perfect right now either.
- The fallback could be important if for example the pod sandbox CRI call failed after a successful container CRI calls. Probably not very common, but it could happen since we do not mark the overall lookups within `parse` as unsuccessful when the container lookups succeeded but the subsequent additional pod sandbox CRI call failed.

**Comment by therealbobo** [comment]
@leogr I did! I also see new commits. Let me check! 👀

**Comment by incertum** [comment]
> @leogr I did! I also see new commits. Let me check! 👀

Only the last commit is new, else conflict-free rebase. The last commit is merely one more renaming and adding more comments pass etc, no changes really.

**Comment by incertum** [comment]
Since we are doing this, let's do it right: I pushed one more commit to really be super consistent also in the order of calls in the CRI code base (plus few more comments). Ok to call me crazy on those details, but as said hopefully we don't need to touch this again for a bit.

**Comment by leogr** [comment]
> Since we are doing this, let's do it right: I pushed one more commit to really be super consistent also in the order of calls in the CRI code base (plus few more comments). Ok to call me crazy on those details, but as said hopefully we don't need to touch this again for a bit.

:+1: Testing this again (in progress).

**Review by leogr — APPROVED** [review]

**Review by FedeDP — APPROVED** [review]
/approve

**Pull request merged**
